feat(forms): add updateOn option to FormBuilder

LayZeeDK opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #19163


## What is the new behavior?
Support `updateOn` option in `FormBuilder` methods `array`, `control` and `group`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
Special care has been taken to keep backwards compatibility with the `extra` object containing optional `validator` and `asyncValidator` properties in the `FormBuilder#group` method so as not to create breaking changes.

Unit tests have been massaged from the form control spec.

LayZeeDK committed:
feat(forms): add updateOn option to FormBuilder

kara commented on packages/forms/src/form_builder.ts:
```diff
@@ -9,7 +9,23 @@
-import {AbstractControl, FormArray, FormControl, FormGroup} from './model';
+import {AbstractControl, AbstractControlOptions, FormArray, FormControl, FormGroup, FormHooks} from './model';
+
+export interface GroupExtra {
```
Comment here about what this is?  For the name, I think `OldFormBuilderOptions` is more descriptive.

LayZeeDK replied:
Seeing that we are currently at version 6.1 and that there will not be another minor release before version 7, do you have any thoughts on adding deprecation notices to these (soon to be) legacy group options?

kara replied:
Hmm, that's a good point. I think yes, but let's do that in a follow-up PR since we need to design a a full deprecation plan.

kara reviewed (commented):
Thanks for the PR! It's looking great, just one comment (also needs a rebase).

JoostK commented on packages/forms/src/form_builder.ts:
```diff
@@ -39,35 +55,65 @@ export class FormBuilder {
+    if (isGroupExtra(extraOrOpts)) {
+      validators = extraOrOpts.validator != null ? extraOrOpts.validator : null;
+      asyncValidators = extraOrOpts.asyncValidator != null ? extraOrOpts.asyncValidator : null;
+    } else if (extraOrOpts != null) {
```
This doesn't seem right. If `isGroupExtra` is satisfied, the `updateOn` property will not be set. Instead of the `GroupExtra` interface I would try to either lift on the usage of `AbstractControlOptions` (which is not public last time I checked) or introduce a public facing interface with all fields being optional. @kara thoughts?

kara replied:
I think it is right as is. We don't want to keep supporting the `GroupExtra` API, so I'd prefer not to add options to it.

JoostK replied:
Agreed, but `AbstractControlOptions` is fully type-compatible with `FormExtra` so there shouldn't be a need for the new interface to be added at all. The bug I described above is real.

kara replied:
I don't think it's type compatible. `asyncValidator` vs `asyncValidators` and `validator` vs `validators`.

JoostK replied:
My bad. I did not spot the additional `s`.

LayZeeDK commented:
Hi @kara,

I renamed the interface and added a description as per your comment.

To my best ability I have rebased from `upstream/master` as well.

kara commented:
@LayZeeDK  Can you take a look at the failing CI test? Looks related to the change: https://travis-ci.org/angular/angular/jobs/418858316.

Also you'll want to lint with `gulp format` to get the linting check green.

LayZeeDK commented:
@kara You are right. I did not catch that the documentaton example was coupled to an end-to-end test. It should be fixed now, but honestly I have no idea since I am having trouble running the tests on Windows.

Additionally, I can never figure out when CI fails because of something I have done.

LayZeeDK commented:
It seems that all that is left is to accept a new golden file for the public API guard.

kara commented:
@LayZeeDK Yes, the last thing needed to get the tests green is to generate a new golden file. You can do so with:

```
bazel run //tools/public_api_guard:forms_forms_api_bin.accept
```

LayZeeDK commented:
Alright, the command seems to bee `bazel run //tools/public_api_guard:forms_forms_api.accept` but fails with this

```
ERROR: **/_bazel_**/go3n4uez/external/rxjs/BUILD.bazel:7:1: Compiling TypeScript (devmode) @rxjs//:lib failed (Exit 1)
external/rxjs/internal/scheduler/AsyncAction.ts:82:12 - error TS1345: An expression of type 'void' cannot be tested for truthiness

82     return clearInterval(id) && undefined || undefined;
              ~~~~~~~~~~~~~~~~~
external/rxjs/internal/scheduler/AsyncAction.ts:82:12 - error TS1345: An expression of type 'void' cannot be tested for truthiness

82     return clearInterval(id) && undefined || undefined;
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

LayZeeDK commented:
Help me out here
```
WARNING: Overflow when watching local filesystem for changes... temporarily falling back to manually checking files for changes
ERROR: /mnt/d/Private/Projects/LayZeeDK/angular/packages/forms/BUILD.bazel:21:1: no such package '@angular_deps//': yarn_install failed: yarn install v1.9.2
info If you think this is a bug, please open a bug report with the information provided in "/home/lars/.cache/bazel/_bazel_lars/c49625842b1d24dd37283e182df182e8/external/angular_deps/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
 (error An unexpected error occurred: "EACCES: permission denied, scandir '/home/lars/.config/yarn/link'".
) and referenced by '//packages/forms:npm_package'
ERROR: Analysis of target '//tools/public_api_guard:forms_forms_api.accept' failed; build aborted: no such package '@angular_deps//': yarn_install failed: yarn install v1.9.2
info If you think this is a bug, please open a bug report with the information provided in "/home/lars/.cache/bazel/_bazel_lars/c49625842b1d24dd37283e182df182e8/external/angular_deps/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
 (error An unexpected error occurred: "EACCES: permission denied, scandir '/home/lars/.config/yarn/link'".
)
INFO: Elapsed time: 1.157s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
FAILED: Build did NOT complete successfully (0 packages loaded)
```

LayZeeDK commented:
Please help me out here
```
WARNING: Overflow when watching local filesystem for changes... temporarily falling back to manually checking files for changes
ERROR: /mnt/d/**/angular/packages/forms/BUILD.bazel:21:1: no such package '@angular_deps//': yarn_install failed: yarn install v1.9.2
info If you think this is a bug, please open a bug report with the information provided in "/home/lars/.cache/bazel/_bazel_lars/c49625842b1d24dd37283e182df182e8/external/angular_deps/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
 (error An unexpected error occurred: "EACCES: permission denied, scandir '/home/lars/.config/yarn/link'".
) and referenced by '//packages/forms:npm_package'
ERROR: Analysis of target '//tools/public_api_guard:forms_forms_api.accept' failed; build aborted: no such package '@angular_deps//': yarn_install failed: yarn install v1.9.2
info If you think this is a bug, please open a bug report with the information provided in "/home/lars/.cache/bazel/_bazel_lars/c49625842b1d24dd37283e182df182e8/external/angular_deps/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
 (error An unexpected error occurred: "EACCES: permission denied, scandir '/home/lars/.config/yarn/link'".
)
INFO: Elapsed time: 1.157s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
FAILED: Build did NOT complete successfully (0 packages loaded)
```

kara commented:
@LayZeeDK Hmm, I'm not sure what's going on there. Maybe a Windows issue with Bazel? (I'm on a Mac, and the command seems to work as expected).  I ran the command for you and added it in a fixup commit so it wouldn't mess with your authorship.

kara commented:
@IgorMinar Assigning to you for public API approval.

LayZeeDK commented:
> I'm not sure what's going on there. Maybe a Windows issue with Bazel?
I tried it both on Bash, MSYS2, and Ubuntu@Windows Subsystem for Linux.

Thank you for helping me out, @kara! I spent more time wrestling build tools than adding code and tests 😅

IgorMinar reviewed (approved):
LGTM for api but let's wait until we branch off 7.0.x before we merge this.

LayZeeDK commented:
@kara Remember a deprecation plan for the legacy form group options in the `FormBuilder`
https://github.com/angular/angular/pull/24599#discussion_r206991872

kara commented:
[presubmit](http://test/OCL:217401600:BASE:217401654:1539730006166:f755997c)

kara commented on packages/forms/src/form_builder.ts:
```diff
@@ -9,7 +9,29 @@
+ */
+export interface LegacyFormGroupOptions {
+  asyncValidator?: AsyncValidatorFn|AsyncValidatorFn[]|null;
+  validator?: ValidatorFn|ValidatorFn[]|null;
```
Actually, after chatting with @IgorMinar , I think we need to remove the `LegacyFormGroupOptions` type altogether and keep the type exactly the same to avoid breaking changes.

LayZeeDK commented:
I need your help to create/roll back the golden file for the API guard, @kara. Thank you 🙂

LayZeeDK commented:
@kara I need your help to modify the golden file for the public API guard again. Thank you! 🙂

JohnyDaison commented:
So does this mean this option will only be available in angular 8, or will it also make it into 7?

LayZeeDK commented:
>           So does this mean this option will only be available in angular 8, or will it also make it into 7?

Hopefully 7.1 based on Igor's comment.

kara committed:
fixup! feat(forms): add updateOn option to FormBuilder

LaljiG commented:
Any firm idea which release this (updateOn via FB group method) will be released in?

LayZeeDK commented:
> Any firm idea which release this (updateOn via FB group method) will be released in?

It was released in `@angular/forms@7.1.0`.
